### PR TITLE
Added ARDUINO_TTGO_LoRa32_V1 board and script for Windows environment

### PR DESCRIPTION
**Pull request by pe1mew**
1. Added ARDUINO_TTGO_LoRa32_V1 board to standard-pinmaps.ino and verified operation to be working OK.
2. Added export batch script for use in Windows environments to extract Arduino compatible library, similar to the shell script that is already available. The batch script will work when no spaces are part of the directory path.

### CODE REVIEW

File: target/arduino/export.bat [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+GOTO :EOF
+
+:DEFAULT_CASE
+   ECHO Unknown command. Type: export.bat -h or --help for options.
+GOTO :EOF
+
+:COPY
+   :: Test if target directory has a terminating back-slash
+   IF NOT "%target_directory:~-1%"=="\" (
+      ECHO [ERROR] Target directory has no terminating "\" (back-slash^). Please correct target directory accordingly.
```
Comment (matthijskooijman) [inline target/arduino/export.bat]: Why is this needed? Below add an extra \ is added after the target directory already in some cases, can this not be always done? Or is it problematic for `COPY` to have a double `\\`?

File: target/arduino/export.bat [hunk 2]
```diff
@@ -0,0 +1,89 @@
+@ECHO off > nul
+:: Batch file to install BasicMAC from the git repository in to an
+:: existing Arduino installation by specifying the location of the 
+:: User library directory e.g.: "C:\Users\UserName\Documents\Arduino\libraries"
+
+:: \name Remko Welling (pe1mew@gmail.com)
+:: \date 8-7-2020
+:: \version 1.0
```
Comment (matthijskooijman) [inline target/arduino/export.bat]: I'd rather not have a date and version in here, those are likely to get outdated and git can already track this better. Having your name as author here is fine, of course.

File: target/arduino/export.bat [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+:: \todo This script cannot work when whitespaces are in the directory path.
+:: As a workaround, export the library else where and move it in to the Arduino library directory.
```
Comment (matthijskooijman) [inline target/arduino/export.bat]: What part breaks exactly? Isn't this something that can be fixed with proper quoting? If not, can the script check for this and show an error message instead?

**Review by matthijskooijman — CHANGES_REQUESTED** [review]
Thanks for the pullrequest. I've picked up your TTGO pinout already (fixed some comments and them manually merged it), so that's in master. Could you maybe rebase this PR on top of that and also remove the extra commits from this PR (probably needs an interactive rebase, so you can remove all commits except for the export script one).

Having an export script for Windows is also nice, thanks for that. I've done a quick review of it, see inline comments and below.

To export the library, you have to pass `--create`, but then the script does not create the directory (but instead insists that it already exists), which seems a bit unexpected to me. Also, copying files into a pre-existing directory is a bit risky, since they it's more likely that you overwrite files (which is why the Linux script creates the directory and explicitly asks to overwrite if it already exists).

It would be nice if the Windows script could work the same as the Linux script, except for `--link` (so not requiring a `--create` option, defaulting to creating a directory and asking to overwrite if it already exists). Not sure how complex that would be, maybe too much work. A simpler alternative could be to:
 - Specify `--create` to create the directory, aborting if it already exists.
 - Specify `--replace` to replace an existing directory (but probably no error if it does not already exists).

Note that with `--replace`, it is probably important to completely *remove* the existing directory completely, rather than just overwriting files inside, in case a file was removed from the library.

File: target/arduino/export.bat (hunk 2 above)
Comment (pe1mew) [reply target/arduino/export.bat]: It's normal behaviour for me to try to keep as much compatible with Doxygen documentation. There is no need to keep versioning here. I'll remove it.

File: target/arduino/export.bat (hunk 3 above)
Comment (pe1mew) [reply target/arduino/export.bat]: I investigated on quoting but it didn't solve the issue. I will investigate on checking for spaces and throw an error. Unless some one comes with a more fundamental solution of coarse.

File: target/arduino/export.bat (hunk 1 above)
Comment (pe1mew) [reply target/arduino/export.bat]: Double slashes are accepted in DOS/Windows. Solved by consistetly using slash in the script.

File: target/arduino/export.bat (hunk 3 above)
Comment (pe1mew) [reply target/arduino/export.bat]: The problem was caused by double quotes that are part of "value" of the value. 
I implemented a solution to remove double quotes from the variable in the batch file. 
The btach file now works with white spaces in the file directory name.

**Comment by pe1mew** [comment]
I have rebased the fork and added the new batch file. A PR will follow.

**Comment by pe1mew** [comment]
Please evaluate changes and Pull when accepted.

**Comment by matthijskooijman** [comment]
> I have rebased the fork and added the new batch file.

Seems you merged in my master branch, rather than rebasing on top of it, so now there's even more commits in the PR than less :-)  Don't worry, though, I can just clean this up easily when merging.

>A PR will follow.

Seems you discovered you can push more commits to updated this PR rather than create a new PR, which is also what I prefer. If you do rebase, you'll have to force push, but that will also work to update the PR (but replaces commits rather than just adding them).

The script installs into a `BasicMAC` subdirectory of the target directory (so you would pass your `Arduino/libraries` directory to the script), but for consistency with the Linux version of the script, I'd rather have it install into the given directory directly (so you'd pass `Arduino/libraries/BasicMAC` to the script). Seems I missed this on my first review, sorry.

You now implemented a "Are you sure" prompt for deleting the old directory. With that, it seems you're already quite close to making this script behave exactly like the Linux version (which would be preferred if possible). That would mean:
 1. If called with `-h` or `--help`, print usage as now (the linux version only knows `--help`, but I think I'll add `-h` there as well).
 2. If called with any other argument starting with `--` print an "unknown option" error message (this one is not so important and can be omitted if needed)
 3. If the directory name passed does not exist, create it and copy files into it directly.
 4. If the directory name passed already exists, show a prompt and if the answer is yes, remove the existing file or directory, create a new directory and copy files into it.

This would mean dropping the `--create` and `--replace` arguments (I proposed the latter mostly since I wasn't sure how complex a "remove existing directory"-prompt would be, but apparently you've figured that one out already).

**Comment by pe1mew** [comment]
Git is not always clear to me, but we will get there eventually. I observed the 9 commits that I intended to have "removed" kind of. 

With respect to your features: you would only like to have one option -h and the flow you described?

**Comment by matthijskooijman** [comment]
> Git is not always clear to me, but we will get there eventually. I observed the 9 commits that I intended to have "removed" kind of.

It's a bit like black magic sometimes :-p

> With respect to your features: you would only like to have one option -h and the flow you described?

Yup, then it will work identically between Linux and Windows (except for the missing `--link` symlink feature, which is probably hard or impossible on Windows).

**Comment by pe1mew** [comment]
Ok, done. The API and behaviour is now compatible with export.sh.

File: target/arduino/export.bat [hunk 4]
```diff
@@ -5,106 +5,101 @@
...
-   ECHO -r  --replace ^<directory^> Replace Arduino library at specified location erasing existing files.
 GOTO :EOF
 
 :DEFAULT_CASE
-   ECHO Unknown command or abort. Type: export.bat -h or --help for options.
+   ECHO Unknown command. Type: export.bat -h or --help for options.
+GOTO :EOF
+
+:RUN
+:: Normal operation.
```
Comment (matthijskooijman) [inline target/arduino/export.bat:43]: I would suggest doing something like `SET target_directory=%1` here, so you can just keep using `target_directory` below (using `option` as the directory name seems a bit confusing IMHO).

File: target/arduino/export.bat [hunk 5]
```diff
@@ -5,106 +5,101 @@
...
+   )
 GOTO :EOF
 
 :DO_REMOVE
-:: remove existing directory, prompt for acknowledge.
-   ECHO Removing directory "%target_directory:"=%\BasicMAC",
+:: Remove existing directory, prompt for acknowledge.
+   ECHO Removing directory "%option:"=%\BasicMAC",
    SET /P AREYOUSURE=Are you sure (Y/[N])?
    IF /I "%AREYOUSURE%" NEQ "Y" GOTO :EOF
```
Comment (matthijskooijman) [inline target/arduino/export.bat:59]: Is this check case sensitive? Would probably be best if both `y` and `Y` work?

File: target/arduino/export.bat [hunk 6]
```diff
@@ -5,106 +5,101 @@
...
+   rmdir /s /q "%option:"=%\BasicMAC"
 
 :DO_CREATE
-:: Copy all files in to the library in the target directory
-   ECHO Ceating Arduino library in: "%target_directory:"=%"
-   IF NOT EXIST "%target_directory:"=%\BasicMAC" (
-      md "%target_directory:"=%\BasicMAC"
+:: Create directory structure for library in target directory
+   ECHO Ceating Arduino library in: "%option:"=%"
+   IF NOT EXIST "%option:"=%\BasicMAC" (
```
Comment (matthijskooijman) [inline target/arduino/export.bat]: This if can be removed, I think, since you just removed the dir if it existed.

**Review by matthijskooijman — COMMENTED** [review]
Looking good! I left some more inline remarks. Also, seems you missed this earlier remark?

> The script installs into a BasicMAC subdirectory of the target directory (so you would pass your Arduino/libraries directory to the script), but for consistency with the Linux version of the script, I'd rather have it install into the given directory directly (so you'd pass Arduino/libraries/BasicMAC to the script). Seems I missed this on my first review, sorry.

File: target/arduino/export.bat (hunk 6 above)
Comment (pe1mew) [reply target/arduino/export.bat:65]: it is not case sensitive

File: target/arduino/export.bat (hunk 4 above)
Comment (pe1mew) [reply target/arduino/export.bat:43]: It's a coding style thing I do support. its done.

File: target/arduino/export.bat (hunk 6 above)
Comment (pe1mew) [reply target/arduino/export.bat]: I want to keep this to ensure that the destination directory structure is capeable to have files copied in to.

Comment (matthijskooijman) [reply target/arduino/export.bat]: I'm not sure I follow that. My suggestion was to remove the if, but leave the `md` (just run it unconditionally)?

**Comment by pe1mew** [comment]
Hi Matthijs, did you finish assessing changes required for PR?

File: target/arduino/export.bat (hunk 6 above)
Comment (matthijskooijman) [reply target/arduino/export.bat]: You resolved this, but have not replied to the question, nor made any changes to resolve this. Can you clarify?

File: target/arduino/export.bat (hunk 5 above)
Comment (matthijskooijman) [reply target/arduino/export.bat:59]: You resolved this, but have not replied to the question, nor made any changes to resolve this. Can you clarify?

**Comment by matthijskooijman** [comment]
I had another look at the script, there's two questions still open above.

Also, looking again, I see that all directories are created and copied explicitly, including e.g. the individual example directories. This means that if e.g. an example is added, the script needs to be modified. Is there a way we can avoid that? i.e. do a recursive copy? Ideally, the individual mkdirs could also be removed when the copy creates each directory it copies to (leaving only a mkdir for the root and `src` directories, like in the bash script).

File: target/arduino/export.bat (hunk 6 above)
Comment (pe1mew) [reply target/arduino/export.bat]:
```
You resolved this, but have not replied to the question, nor made any changes to resolve this. Can you clarify?
```
I do not understad what you are referring to. At July 16 I stated that I want to keep the confirmation.
Probably that a commit from my side was not executed nor observed by you. who can tell? I'll try to update again today.

**Comment by pe1mew** [comment]
````
I had another look at the script, there are two questions still open above.
````
Great, it took a while. 
````
Also, looking again, I see that all directories are created and copied explicitly, including e.g. the individual example directories. This means that if e.g. an example is added, the script needs to be modified. Is there a way we can avoid that? i.e. do a recursive copy? Ideally, the individual mkdirs could also be removed when the copy creates each directory it copies to (leaving only a mkdir for the root and src directories, like in the bash script).
````

- I have implemented a copy of the example directories as a whole.
- The Y/N confirmation is not case sensitve.

I will close the PR, commit the changes and do another PR.

File: target/arduino/export.bat (hunk 6 above)
Comment (matthijskooijman) [reply target/arduino/export.bat]: You have not replied to my comment from july 16th:

 > I'm not sure I follow that. My suggestion was to remove the if, but leave the md (just run it unconditionally)?

Comment (matthijskooijman) [reply target/arduino/export.bat]: You say you want to "keep the confirmation", but what confirmation do you mean exactly? Is the `IF NOT EXIST "%option:"=%\BasicMAC" (` the confirmation you mean? Or do you mean that `      md "%target_directory:"=%\BasicMAC"` is the confirmation? Since I'm proposing to remove the `if`, but keep the `md` (unless changes to the copy command could make the `md` unneeded, that I'd also prefer to remove those, but that's a separate issue).

**Comment by matthijskooijman** [comment]
@pe1mew I noticed a few more things where your script behaved differently (i.e. `export.sh` expects the library directory itself to be passed, while your script expected the parent `libraries` directory). Rather than bouncing forward and back, I fired up Windows and played around with the script myself to fix the remaining incompatibilities, but ended up pretty much rewriting the script completely to follow the same interface and structure as the original `export.sh`.

The result in is #20, maybe you could see if that version of the batch script works for you and if it looks good?

**Comment by matthijskooijman** [comment]
I'm closing this, now that #20 has been merged.
